exp/metric-gen/main.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/genall/help"
+	prettyhelp "sigs.k8s.io/controller-tools/pkg/genall/help/pretty"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"k8s.io/kube-state-metrics/exp/metric-gen/metric"
+)
+
+var (
+	allGenerators = map[string]genall.Generator{
+		"metric": metric.Generator{},
+	}
+
+	allOutputRules = map[string]genall.OutputRule{
+		"dir":       genall.OutputToDirectory(""),
+		"none":      genall.OutputToNothing,
+		"stdout":    genall.OutputToStdout,
+		"artifacts": genall.OutputArtifacts{},
+	}
+
+	// optionsRegistry contains all the marker definitions used to process command line options
+	optionsRegistry = &markers.Registry{}
+)
+
+func init() {
+	for genName, gen := range allGenerators {
```

Comment:
In my opinion the controller-gen way of handling input flags is totally overengineered for our use case (assuming we just need a few flags).

I would try to get rid of all of this complexity

I ~ did this with the existing crd generator here: https://github.com/kubernetes-sigs/cluster-api/blob/296224d73b33a3619e681c6c87a2dc3c694de401/hack/tools/runtime-openapi-gen-2/main.go#L80-L92

(the code is not very cleaned up, but it might show that you don't need most of the stuff your using here like genall, ...)

Do you need cobra or just a few flags? then the go flag package might be the easiest way.